File: src/service_example/service_server/src/service_server.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR. All Rights Reserved.
 * SPDX-License-Identifier: LicenseRef-Segar-Proprietary
 *
 * PROPRIETARY AND CONFIDENTIAL. See ./LICENSE
 * for license terms and restrictions.
 *****************************************************************************/

#include <memory>
#include <string>

#include "example/srv/SetCameraInfo.hpp"

#include "segar/segar.h"

int main(int argc, char* argv[]) {
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);

  auto node = rti::segar::CreateNode("set_camera_info_server");
  RETURN_VAL_IF(!node, EXIT_FAILURE);

  using SetCameraInfo = example::srv::SetCameraInfo;
  auto callback = [](const std::shared_ptr<SetCameraInfo::Request>& request,
                     std::shared_ptr<SetCameraInfo::Response>& response) {
    response->success(true);
    response->status_message("Camera info set successfully");
    AINFO << "Request camera width: " << request->camera_info().width()
          << ", Response msg:" << response->status_message();
  };
  auto service =
      node->CreateService<SetCameraInfo>("set_camera_info", callback);
  RETURN_VAL_IF(!service, EXIT_FAILURE);

  AINFO << "Waiting for requests...";
  rti::segar::WaitForShutdown();

  return EXIT_SUCCESS;
}

File: src_python/service_example/service_server/src/service_server.py
#!/usr/bin/env python3

from example.srv import SetCameraInfo
from segar.python import segar


def main():
    if not segar.Init("set_camera_info_server"):
        return 1
    node = segar.CreateNode("set_camera_info_server")
    if node is None:
        return 1

    def on_request(request):
        print(f"[service_server] request width={request.camera_info.width}")
        return SetCameraInfo.Response(
            success=True,
            status_message="Camera info set successfully",
        )

    service = node.CreateService("set_camera_info", SetCameraInfo, on_request)
    if service is None:
        print("[service_server] create service failed")
        return 1

    print("[service_server] waiting for requests...")
    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
